ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
-
-  char *mkey = StringValuePtr(rb_key);
-  char *mvalue = StringValuePtr(rb_value);
-
-  memcached_return_t rc = memcached_replace(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(rb_ttl), NUM2INT(rb_flags));
-  handle_memcached_return(rc);
-  return (rc == MEMCACHED_SUCCESS);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_replace(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+	rb_memcached_error_check(rc);
+	return (rc == MEMCACHED_SUCCESS) ? Qtrue : Qfalse;
 }
 
 static VALUE
 rb_connection_prepend(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(rb_key);
-  char *mvalue = StringValuePtr(rb_value);
-
-  memcached_return_t rc = memcached_prepend(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(rb_ttl), NUM2INT(rb_flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_prepend(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_append(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(rb_key);
-  char *mvalue = StringValuePtr(rb_value);
-
-  memcached_return_t rc = memcached_append(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(rb_ttl), NUM2INT(rb_flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_append(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+	rb_memcached_return(rc);
 }
 
 static VALUE
 rb_connection_set_prefix(VALUE self, VALUE rb_prefix)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  memcached_return_t rc;
-  if(Qnil == rb_prefix) {
-    rc = memcached_callback_set(ctx->memc, MEMCACHED_CALLBACK_NAMESPACE, NULL);
-  } else {
-    char *prefix = StringValuePtr(rb_prefix);
-    rc = memcached_callback_set(ctx->memc, MEMCACHED_CALLBACK_NAMESPACE, prefix);
-  }
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+	const char *prefix = NULL;
+
+	UnwrapMemcached(self, mc);
+
+	if (!NIL_P(rb_prefix)) {
+		Check_Type(rb_prefix, T_STRING);
+		prefix = StringValueCStr(rb_prefix);
```

Comment:
`StringValueCStr` ensures that the returning string is NULL terminated, has no NULL bytes at all (it will raise if it finds a NULL byte, as opposed to `StringValuePtr`, which will return a pointer with NULL bytes and lead to weird behavior). It's what we want for setting the connection's prefix.